staging/src/k8s.io/apiserver/pkg/storage/cacher/btree_store.go

```diff
@@ -0,0 +1,215 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/btree"
+	"k8s.io/client-go/tools/cache"
+)
+
+type btreeIndexer interface {
+	cache.Indexer
+	Clone() btreeIndexer
+}
+
+type btreeStore struct {
+	lock sync.RWMutex
+	tree *btree.BTree
+}
+
+func newBtreeStore(degree int) *btreeStore {
+	return &btreeStore{
+		tree: btree.New(degree),
+	}
+}
+
+func (t *btreeStore) Add(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.addOrUpdateLocked(obj)
+}
+
+func (t *btreeStore) Update(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.addOrUpdateLocked(obj)
+}
+
+func (t *btreeStore) Delete(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	storeElem, ok := obj.(storeElement)
+	if !ok {
+		return fmt.Errorf("obj not a storeElement: %#v", obj)
+	}
+	item := t.tree.Delete(storeElem)
+	if item == nil {
+		return fmt.Errorf("obj does not exist")
+	}
+
+	return nil
+}
+
+func (t *btreeStore) List() []interface{} {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	items := make([]interface{}, 0, t.tree.Len())
+	t.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(interface{}))
+		return true
+	})
+
+	return items
+}
+
+func (t *btreeStore) ListKeys() []string {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	items := make([]string, 0, t.tree.Len())
+	t.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(storeElement).Key)
+		return true
+	})
+
+	return items
+}
+
+func (t *btreeStore) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	storeElem, ok := obj.(storeElement)
+	if !ok {
+		return nil, false, fmt.Errorf("obj is not a storeElement")
+	}
+	item = t.tree.Get(storeElem)
+	if item == nil {
+		return nil, false, nil
+	}
+
+	return item, false, nil
+}
+
+func (t *btreeStore) GetByKey(key string) (item interface{}, exists bool, err error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	t.tree.Ascend(func(i btree.Item) bool {
+		if key == i.(storeElement).Key {
+			item = i
+			exists = true
+			return false
+		}
+		return true
+	})
+
+	return item, exists, nil
+}
+
+func (t *btreeStore) Replace(objs []interface{}, _ string) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	t.tree.Clear(false)
+	for _, obj := range objs {
+		err := t.addOrUpdateLocked(obj)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *btreeStore) Resync() error {
+	// Nothing to do.
+	return nil
+}
+
+// TODO(MadhavJivrajani): This is un-implemented for now. Stubbing this out
```

Comment:
In general you would need to reimplement something similar to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go#L135-L323

However - our case is somewhat special, as there is exactly one index defined and only for pods.
So we can probably shortcut it and:
- implement just a single index using a btree [it will be btree with key="index value" and value="set of keys matching that value"]
- use that here (we can have a strict validation that will fail creation if there is more than 1 indexer)
[we can eventually generalize it, but it's not needed now]

Reply:
I actually thought about that and in the KEP I proposed to ignored indexes in general:
https://github.com/kubernetes/enhancements/pull/3274

PTAL

Reply:
Ack
Thanks @wojtek-t! Will get back to this post release :)

Reply:
As I described here:
https://github.com/kubernetes/enhancements/pull/3274/files#diff-3d93eb20b3400b8a937e7ab25e0aca359f73d8600271605df260b3b83bad06a7R301-R303

let make that very simple:
(1) if we're at "now" - let's have the index implemented the same way as we have now:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go#L70
[so basically something like:
```
store.lock()
current := store.btree.Clone()
indexed := store.index[myvalue] (probably deep-copied, depending on the implementation)
store.unlock()

// procee the request using "current" and "indexed"
```
(2) let's ignore that for continuation
[index is just an optimization, and for continuation, we can just go ahead and process all items